Support for an advanced “persistentCache”

martinheidegger opened the pull request:
There are a few implementations of persistent caches for browserify such as [`persistify`](https://github.com/royriojas/persistify) or [`browserify-incremental`](https://github.com/jsdf/browserify-incremental). But looking at each implementation it seemed like  they struggle with the interface given by `module-deps` to cache: which is just an object lookup. This PR adds an additional cache interface to `module-deps` that should make it possible to implement not just resource-efficient but also less hacky caching solutions for browserify.

terinjokes commented on readme.markdown:
```diff
@@ -94,6 +94,38 @@ contents for browser fields, main entries, and transforms
+        }
+        if (isInCache()) {
+            return cb(null, {
+                source: fs.readFileSync(file, 'utf8'), // String of the fully processed file
```
Since this is the "in cache" case, the example should probably show the contents coming from somewhere other than `fs`.

martinheidegger replied:
It does now.

terinjokes commented on readme.markdown:
```diff
@@ -94,6 +94,38 @@ contents for browser fields, main entries, and transforms
+        // optional (can be null) stream that provides the data from
+        // the hard disk, can be provided in case the file data is used
+        // to evaluate the cache identifier
+        stream = fs.createReadStream(file)
```
Can there be a better example here, since this is what I expect `fallback` to be doing.

martinheidegger replied:
I updated the example. it now uses a db.

terinjokes commented on index.js:
```diff
@@ -376,19 +381,41 @@ Deps.prototype.walk = function (id, parent, cb) {
-                fromSource(file, body.toString('utf8'), pkg, fakePath);
-            }))
-        ;
+        self.persistentCache(file, id, pkg, function fallback (stream, cb) {
```
Is it possible to move the fallback function somewhere other than an inline parameter here?

martinheidegger replied:
Moved it to an outside scope.

martinheidegger commented:
@terinjokes Thank you for your review! I updated the PR. One thing to mention: I changed the implementation. Now the fallback takes an optional string instead of a stream. This change might seem odd, I came to the conclusion that this would be the better API when I tried to implement a cache. The cache _could_ read the source in order to create a Key but if it did that with streams (i.e. to create a hash) then it would need to keep all the file data in memory and create yet another stream off of the in-memory data which would just make the impl. more complex without a real win for the module-deps.

martinheidegger commented:
@terinjokes Thank you for approving the PR, how good are the chances for a merge?

terinjokes commented:
I'd prefer to have someone else take a look first.

martinheidegger commented:
Whom would be good to ask?

martinheidegger commented:
Looked into the `npm owners` and pinging @substack @zertosh @jmm @MellowMelon @ahdinosaur for a review.

ahdinosaur commented on index.js:
```diff
@@ -550,6 +582,13 @@ function xhas (obj) {
     return true;
 }
 
+function toStream (dataAsString) {
```
nice refactor :+1:

ahdinosaur commented on index.js:
```diff
@@ -376,19 +378,49 @@ Deps.prototype.walk = function (id, parent, cb) {
+        }
     });
 
+    function getDeps (file, src) {
```
more goods :rabbit:

ahdinosaur commented on index.js:
```diff
@@ -27,6 +27,11 @@ function Deps (opts) {
     
     this.basedir = opts.basedir || process.cwd();
+    this.persistentCache = opts.persistentCache || function (file, id, pkg, fallback, cb) {
+        setImmediate(function () {
```
why `setImmediate` here?

martinheidegger replied:
͓̞̦̼͎. ̖̱̗̝͉͈Z̧̬͎̥͕̻̜AḺ̡̼G̝͓͘O̴̫̹̰̯͕!͈̱̟͈̕

ahdinosaur replied:
of course, thanks!

ahdinosaur commented on index.js:
```diff
@@ -376,19 +378,49 @@ Deps.prototype.walk = function (id, parent, cb) {
+
+        function persistentCacheFallback (dataAsString, cb) {
+            var stream
+            if (dataAsString) {
```
ternary? also semicolon styles missing here;

ahdinosaur commented on index.js:
```diff
@@ -376,19 +378,49 @@ Deps.prototype.walk = function (id, parent, cb) {
+                            source: src,
+                            package: pkg,
+                            deps: deps.reduce(function (deps, dep) {
+                                deps[dep] = true
```
if (pedantic);

ahdinosaur commented on readme.markdown:
```diff
@@ -94,6 +94,50 @@ contents for browser fields, main entries, and transforms
+        // then you can pass `null` instead of the source and the fallback will
+        // fetch the data by itself.
+        //
+        fallback(fileData, function (error, cacheableEntry) {
```
where did `fallback` come from here? is missing from above description of function arguments.

ahdinosaur commented on test/cache_persistent.js:
```diff
@@ -0,0 +1,81 @@
+    p.on('error', function (err) { t.equals(err.message, 'foo') });
+});
+
+test('allow passing of a different stream', function (t) {
```
what stream? do you mean a different string to `fallback`?

martinheidegger replied:
Yes, it used to be a stream :)

martinheidegger commented:
Thanks for the review @ahdinosaur! Pushed a few fixes to the PR in regards to the linting.

martinheidegger commented:
Two approved reviews 😃

martinheidegger commented:
@terinjokes @ahdinosaur Are the chances of a merge higher now with two PR's? 😅

martinheidegger commented:
Note: I published an implementation of persistentCache here: https://github.com/martinheidegger/browserify-persist-fs

substack commented:
We should also ping @jsdf and @royriojas upstream to make sure they know about this patch so they can make the best use of these changes.

ahdinosaur commented:
@martinheidegger hey, so the persistent cache breaks when using [`bulk-require`](https://github.com/substack/bulk-require) and [`bulkify`](https://github.com/substack/bulkify). specifically when you create a new file in your bulk require glob, it is ignored and built as before. i see you tried to fix file events with https://github.com/substack/module-deps/pull/127, but as far as i can tell it's still broken when using that fix.
